Cython/Utility/ModuleSetupCode.c

```diff
@@ -2484,3 +2488,259 @@ static int __Pyx_VersionSanityCheck(void) {
 #else
 #define __Pyx_shared_in_cpython_freethreading(x)
 #endif
+
+////////////////////////// MultiPhaseInitModuleState.proto /////////////
+
+#if CYTHON_PEP489_MULTI_PHASE_INIT && CYTHON_USE_MODULE_STATE
+// This defines an ad-hoc, single module version of PyState_FindModule that
+// works for multi-phase init modules. It's intended to be the last option
+// when all the other official ways of getting the module are unavailble.
+static PyObject *__Pyx_State_FindModule(void*); /* proto */
+static int __Pyx_State_AddModule(PyObject* module, void*); /* proto */
+static int __Pyx_State_RemoveModule(void*); /* proto */
+#elif CYTHON_USE_MODULE_STATE
+#define __Pyx_State_FindModule PyState_FindModule
+#define __Pyx_State_AddModule PyState_AddModule
+#define __Pyx_State_RemoveModule PyState_RemoveModule
+#endif
+
+////////////////////////// MultiPhaseInitModuleState /////////////
+
+#if CYTHON_PEP489_MULTI_PHASE_INIT && CYTHON_USE_MODULE_STATE
+
+#ifndef CYTHON_MODULE_STATE_LOOKUP_THREAD_SAFE
+// If you're using multiple interpreters but a single GIL then
+// this can be undefined for a bit of speed.
+// Isolated subinterpreters were added in 3.12, and nogil in 3.13, so before that
+// we can safely assume that we're protected by the GIL.
+// TODO - turn this off as appropriate when the user is able to set
+// Py_MOD_PER_INTERPRETER_GIL_SUPPORTED explicitly.
+#if (CYTHON_COMPILING_IN_LIMITED_API || PY_VERSION_HEX >= 0x030C0000)
+  // Numbers greater 1 can be used to force an implementation
+  #define CYTHON_MODULE_STATE_LOOKUP_THREAD_SAFE 1
+#else
+  #define CYTHON_MODULE_STATE_LOOKUP_THREAD_SAFE 0
+#endif
+#endif
+
+// In order of preference 
+// - use a shared lock if available (from language std library if possible)
+// - use PyMutex
+// - use a language standard library non-shared mutex
+// - PyThread_lock isn't acceptable since we can't initialize it in a thread safe way
```

Comment:
Came up with. The implementation of `PyState_AddModule` in Python isn't too similar. Each single-phase module object has an attribute that stores that module index, and the interpreter state has a table of module indices. So it requires structure in Python to book-keep it. I don't think that applies to multiphase

The order of preference is largely based on 
* strongly prefer a lock that allows many simultaneous reads. But any lock will work.
* weakly prefer things that are in Python.h, otherwise prefer things that are in the language standard library, otherwise whatever else is around.